core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1364,6 +1385,16 @@ private[spark] class DAGScheduler(
                   if (job.numFinished == job.numPartitions) {
                     markStageAsFinished(resultStage)
                     cleanupStateForJobAndIndependentStages(job)
+                    try { // cancelTasks will fail if a SchedulerBackend does not implement killTask
+                      logInfo(
+                        s"Job ${job.jobId} is finished. Killing speculative tasks for this job")
+                      // ResultStage is only used by this job. It's safe to kill speculative or
+                      // zombie tasks in this stage.
+                      taskScheduler.cancelTasks(stageId, shouldInterruptTaskThread(job))
```

Comment:
Thanks, @zsxwing . This looks promising to reduce the flakiness in some test suite.

Reply:
IIRC `cancelTasks()` will fail the stage (maybe it's okay here coz the stage has been marked completed), if we just want to kill speculative/zombie tasks then maybe we shall call `killAllTaskAttempts()` ?